hbase-server/src/main/java/org/apache/hadoop/hbase/security/provider/OAuthBearerSaslServerAuthenticationProvider.java

```diff
@@ -0,0 +1,99 @@
+ */
+package org.apache.hadoop.hbase.security.provider;
+
+import java.io.IOException;
+import java.security.PrivilegedExceptionAction;
+import java.util.Map;
+import javax.security.sasl.Sasl;
+import javax.security.sasl.SaslException;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hbase.security.auth.AuthenticateCallbackHandler;
+import org.apache.hadoop.hbase.security.oauthbearer.internals.OAuthBearerSaslServerProvider;
+import org.apache.hadoop.hbase.security.oauthbearer.internals.knox.OAuthBearerSignedJwtValidatorCallbackHandler;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.hadoop.security.token.SecretManager;
+import org.apache.hadoop.security.token.TokenIdentifier;
+import org.apache.yetus.audience.InterfaceAudience;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@InterfaceAudience.Private
+public class OAuthBearerSaslServerAuthenticationProvider
+    extends OAuthBearerSaslAuthenticationProvider
+    implements SaslServerAuthenticationProvider {
+
+  private static final Logger LOG = LoggerFactory.getLogger(
+    OAuthBearerSaslServerAuthenticationProvider.class);
+  private Configuration hbaseConfiguration;
+  private boolean initialized = false;
+
+  static {
+    OAuthBearerSaslServerProvider.initialize(); // not part of public API
+    LOG.info("OAuthBearer SASL server provider has been initialized");
+  }
+
+  @Override public void init(Configuration conf) throws IOException {
+    this.hbaseConfiguration = conf;
+    this.initialized = true;
+  }
+
+  @Override public AttemptingUserProvidingSaslServer createServer(
+    SecretManager<TokenIdentifier> secretManager, Map<String, String> saslProps)
+    throws IOException {
+
+    if (!initialized) {
+      throw new RuntimeException(
+        "OAuthBearerSaslServerAuthenticationProvider must be initialized first.");
+    }
+
+    UserGroupInformation current = UserGroupInformation.getCurrentUser();
+    String fullName = current.getUserName();
+    LOG.debug("Server's OAuthBearer user name is {}", fullName);
+    LOG.debug("OAuthBearer saslProps = {}", saslProps);
+
+    try {
+      return current.doAs(new PrivilegedExceptionAction<AttemptingUserProvidingSaslServer>() {
+        @Override
+        public AttemptingUserProvidingSaslServer run() throws SaslException {
+          AuthenticateCallbackHandler callbackHandler =
+            new OAuthBearerSignedJwtValidatorCallbackHandler();
+          callbackHandler.configure(hbaseConfiguration, getSaslAuthMethod().getSaslMechanism(),
+            saslProps);
+          return new AttemptingUserProvidingSaslServer(Sasl.createSaslServer(
+            getSaslAuthMethod().getSaslMechanism(), null, null, saslProps,
+            callbackHandler), () -> null);
+        }
+      });
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Failed to construct OAUTHBEARER SASL server");
```

Comment:
Please use another exception type.

Reply:
Do you have a specific recommendation? Why not RuntimeException?